hugo/static/cgi-bin/watermark.cgi

```diff
@@ -0,0 +1,166 @@
+    tmp_dir = tempfile.mkdtemp(prefix='wmk_')
+    input_pdf = Path(tmp_dir) / 'input.pdf'
+    output_pdf = Path(tmp_dir) / 'output.pdf'
+
+    with open(input_pdf, 'wb') as f:
+        # stream copy to avoid loading entire file in memory
+        chunked = 0
+        while True:
+            buf = pdf_item.file.read(64 * 1024)
+            if not buf:
+                break
+            chunked += len(buf)
+            if chunked > MAX_BYTES:
+                f.close()
+                http_error(400, 'File exceeded size limit during upload.')
+            f.write(buf)
+
+    # Validate PDF magic
+    try:
+        with open(input_pdf, 'rb') as f:
+            head = f.read(8)
+            if b'%PDF-' not in head:
+                http_error(400, 'Uploaded file is not a PDF.')
+    except Exception:
+        http_error(400, 'Could not read uploaded PDF.')
+
+    add_footer = find_add_footer()
+    if not add_footer.exists():
+        http_error(500, 'Server configuration error: add_footer.py not found.')
+
+    cmd = [sys.executable, str(add_footer)]
+    if page_start is not None:
+        cmd += ['-p', str(page_start)]
+    # Convert embedded newlines are preserved; add_footer.py will handle them
+    cmd += [str(input_pdf), str(output_pdf), footer_text]
+
+    try:
+        proc = subprocess.run(cmd, capture_output=True, text=True, timeout=40)
+    except subprocess.TimeoutExpired:
+        http_error(500, 'Processing timed out.')
+
+    if proc.returncode != 0 or not output_pdf.exists():
+        stderr = (proc.stderr or '').strip()
+        stdout = (proc.stdout or '').strip()
+        if debug_mode:
+            # Return full diagnostic (no truncation) for troubleshooting.
+            diag = [
+                'Status: processing failed',
+                'Command: ' + ' '.join(cmd),
+                f'Return code: {proc.returncode}',
+                '--- STDERR ---', stderr or '<empty>',
+                '--- STDOUT ---', stdout or '<empty>'
+            ]
+            http_error(500, '\n'.join(diag))
+        else:
+            merged = stderr or stdout or 'Unknown error'
+            http_error(500, f'Failed to process PDF. (Add ?debug=1 for details)\n{merged[:400]}')
+
+    # Success: stream file
+    size = output_pdf.stat().st_size
+    # Some Apache configurations are picky about the "Status" header in CGI output.
+    # Emit only standard headers followed by a blank line, then raw PDF bytes.
+    sys.stdout.write('Content-Type: application/pdf\r\n')
+    sys.stdout.write('X-Content-Type-Options: nosniff\r\n')
+    sys.stdout.write(f'Content-Disposition: attachment; filename="watermarked.pdf"\r\n')
+    sys.stdout.write(f'Content-Length: {size}\r\n')
+    sys.stdout.write('\r\n')
+    sys.stdout.flush()
+    with open(output_pdf, 'rb') as f:
+        shutil.copyfileobj(f, sys.stdout.buffer)
+
+    # Cleanup temp dir
+    try:
+        shutil.rmtree(tmp_dir)
+    except Exception:
+        pass
+
```

Comment:
If any error is raised before, this is never reached, potentially filling up the disk with temporary files. You probably want to wrap the entire main function in a try/except block (and give it the temporary directory as an argument).